node/src/ledger/ledger_service.rs

```diff
@@ -430,79 +429,49 @@ impl LedgerCtx {
         Ok(computed_hash)
     }
 
-    // TODO(tizoc): Only used for the current workaround to make staged ledger
-    // reconstruction async, can be removed when the ledger services are made async
-    pub fn staged_ledger_reconstruct_result_store(&mut self, staged_ledger_hash: LedgerHash) {
-        let staged_ledger =
-            std::mem::take(&mut *self.sync.staged_ledger_just_reconstructed.lock().unwrap())
-                .unwrap();
-
-        openmina_core::debug!(openmina_core::log::system_time();
-            kind = "LedgerService::staged_ledger_reconstruct_result_store",
-            summary = "Storing just reconstructed staged ledger",
-            staged_ledger_hash = staged_ledger_hash.to_string(),
-        );
-
-        self.sync
-            .staged_ledgers
-            .insert(staged_ledger_hash, staged_ledger);
-    }
-
     pub fn staged_ledger_reconstruct(
         &mut self,
         snarked_ledger_hash: LedgerHash,
         parts: Option<Arc<StagedLedgerAuxAndPendingCoinbasesValid>>,
-    ) {
+    ) -> Result<LedgerHash, String> {
         let staged_ledger_hash = parts
             .as_ref()
             .map(|p| p.staged_ledger_hash.clone())
             .unwrap_or_else(|| snarked_ledger_hash.clone());
         let snarked_ledger = self.sync.snarked_ledger_mut(snarked_ledger_hash.clone());
         let mask = snarked_ledger.copy();
 
-        let event_sender = self
-            .event_sender
-            .clone()
-            .expect("Must set an event sender in the ledger context");
-
-        let staged_ledger_just_reconstructed = self.sync.staged_ledger_just_reconstructed.clone();
-
-        // TODO(tizoc): Workaround to make staged ledger
-        // reconstruction async, can be removed when the ledger services are made async
-        std::thread::spawn(move || {
-            let result = if let Some(parts) = parts {
-                let states = parts
-                    .needed_blocks
-                    .iter()
-                    .map(|state| (state.hash().to_fp().unwrap(), state.clone()))
-                    .collect::<BTreeMap<_, _>>();
-
-                StagedLedger::of_scan_state_pending_coinbases_and_snarked_ledger(
-                    (),
-                    &CONSTRAINT_CONSTANTS,
-                    Verifier,
-                    (&parts.scan_state).into(),
-                    mask,
-                    LocalState::empty(),
-                    parts.staged_ledger_hash.0.to_field(),
-                    (&parts.pending_coinbase).into(),
-                    |key| states.get(&key).cloned().unwrap(),
-                )
-            } else {
-                StagedLedger::create_exn(CONSTRAINT_CONSTANTS.clone(), mask)
-            };
+        let result = if let Some(parts) = parts {
```

Comment:
@Sventimir something that was useful about the previous workaround was that while the staged ledger was being reconstructed (the most expensive ledger operation) other requests could still be served.

Unless it gets complicated it would be good to recover that aspect.

An important detail is that in general, usually when ledgers are being mutated, it is because they are being constructed as a new `Mask` (so whoever is mutating a ledger will have exclusive access to it, because until the ledger construction is complete, nobody will need to read from it). And since `Mask`s can be sent across threads, spawning a new thread to handle the construction task should not be an issue (this is why it worked before).

It is not super important for this to be changed now, because @binier is already solving it in his refactoring (in general, not just for this call) by separating read/construct operations based on what I described in the previous paragraph.

Reply:
I believe it should now work as you want.